### DISCUSSION TITLE
Extract subtitles from a very strange file

### DISCUSSION THREAD

**Discussion by Shake34** [SubtitleEdit/subtitleedit#7037] [Q&A]
I'm trying to extract subtitles from a very strange file. I am attaching a small example of these videos.
I tried using Subtitle Edit, unfortunately it says it can't find subtitles in the file.

It's very strange how these subtitles are embedded in the file.  On the one hand they look like hardcode captions, because no program recognizes them as subtitles. On the other hand, I can deactivate and activate them via Vlc. 
However, if I try to extract them as hardcore captions, they are not seen by any program. Similarly, if I try to extract them as normal subtitles, they are not recognized.

https://github.com/SubtitleEdit/subtitleedit/assets/137782699/46a64432-de99-4668-83b4-aee6d819993c

**Comment by jmaraujouy** [comment]
Altough SE cannot open the subtitles, it seems that _CCExtractGUI_ can extract them. It's a SRT file.

But, I don't know if your short clip is corrupted, because they subtitles seems corrupted as well.

[video.zip](https://github.com/SubtitleEdit/subtitleedit/files/11878119/video.zip)

```
1
00:00:01,667 --> 00:00:06,671
pa                              
   arrticul oly did,n thisgg    
 suthested that he thought at   

2
00:00:06,739 --> 00:00:10,041
    wathere pas one ulicarmp    
 coinonent ir the awa that s    
```


```
Opening file: video.mp4
Detected MP4 box with name: ftyp
Detected MP4 box with name: moov

File seems to be a MP4

Analyzing data with GPAC (MP4 library)
Opening 'video.mp4': ok
Track 1, type=vide subtype=avc1
Track 2, type=soun subtype=MPEG
MP4: found 2 tracks: 1 avc and 0 cc
Processing track 1, type=vide subtype=avc1
Processing track 2, type=soun subtype=MPEG

Closing media: ok
Found 1 AVC track(s). Found no dedicated CC track(s).


Total frames time:	  00:00:10:043  (301 frames at 29,97fps)

Min PTS:				00:00:00:000
Max PTS:				00:00:10:142
Length:				 00:00:10:142

Done, processing time = 0 seconds
Issues? Open a ticket here
https://github.com/CCExtractor/ccextractor/issues
```

**Comment by GrampaWildWilly** [comment]
I am attaching 2 screenshots of this video playing in VLC.  It's just 10 seconds long & there are only 2 closed captions in that short time.  You'll notice that the captions VLC is able to display just fine bear no resemblance to what you extracted as an srt.  This tells me that your application in fact did not successfully extract the captions.  Your application actually corrupted the captions.

Also, correct me if I'm wrong.  Closed captions are not srt, vtt, or any type other than closed captions.  Any program that extracts closed captions will extract them as whatever type you tell it to use.  You told your program to extract these closed captions as type srt, so that's what you got.  If you told your application to extract them as vtt, that's what it would do.

![Caption #1](https://github.com/SubtitleEdit/subtitleedit/assets/75588964/7c87bb6e-8915-4fb0-b080-15be2f030f94)

![Caption #2](https://github.com/SubtitleEdit/subtitleedit/assets/75588964/6eedf6dc-fef6-4499-8ac0-ae887bc2fc92)

**Comment by GrampaWildWilly** [comment]
While that video is playing, the VLC Media Information shows what's in the image I'm attaching to this post.  Using the VLC v command, I tried all the subtitle tracks.  Closed captions 2, 3, & 4 were empty.  Only closed captions 1 contained any subtitles, as you can see in the 2 images I just posted above.

<img width="316" alt="VLC Media Information" src="https://github.com/SubtitleEdit/subtitleedit/assets/75588964/30bae6a2-775d-445a-b2a4-48ba45145181">

**Comment by jmaraujouy** [comment]
> Also, correct me if I'm wrong. Closed captions are not srt, vtt, or any type other than closed captions. Any program that extracts closed captions will extract them as whatever type you tell it to use. You told you program to extract these closed captions as type srt, so that's what you got.

You are not wrong. 😂 I haven't used that program in a long time and I was confusing it with another one I use often (_gMKVExtractGUI_).

But I didn't use the latest version. Maybe you can try it, and perhaps the issue with "garbage characters" in the extracted subtitle is fixed in the latest version.

For my test, I used v0.88. Latest version is v0.94. Good luck. 😜

https://github.com/CCExtractor/ccextractor

**Comment by GrampaWildWilly** [comment]
I ran ffprobe on the video.  The results are attached.

I tried about 6 different applications I found through a web search.  I did find CCExtractor but I purposely avoided that because you had already tried it.  I installed each app I found & tried to get it to extract the closed captions from this video.  Not a single one of them recognized any captions.

OK.  I'll try this CCExtractor thing.  Stay tuned.

[ffprobe.txt](https://github.com/SubtitleEdit/subtitleedit/files/11880004/ffprobe.txt)

**Comment by GrampaWildWilly** [comment]
I got 0.94.  It has an utterly pathological GUI.  Alt+Spacebar does not open the main context menu.  You have to explicitly click the mouse in the upper left corner on the icon.  When you select Size, it will allow you to drag either side left & right.  But you can't drag either the top or the bottom.  This is important because the bottom of nearly every screen is cut off, & the pages don't scroll.  Even Maximize doesn't pull the bits out of sight into view.  So I don't know what I'm missing down there.

In any case, I tried targeting the output to srt & then vtt.  I got empty files.  Well, the vtt started with the standard line WEBVTT, but it was empty after that.

I even tried to tell ffmpeg to extract the captions to a vtt.  This is what I got:

[out#0/webvtt @ 00000000004a76c0] Output file does not contain any stream

Adding -map 0:0:s got me this:

[mov,mp4,m4a,3gp,3g2,mj2 @ 00000000003ee0c0] Invalid stream specifier: 0:s.
    Last message repeated 3 times
Stream map '0:0:s' matches no streams.

-map 0:s also failed.  In other words, ffmpeg can't detect the closed captions.

VLC seems to be the only program that can read these closed captions.  It's very strange.

**Comment by niksedk** [comment]
Interesting file, and yes, only VLC seems to read this correctly.

SE does not have any EIA-608 decoding, and I do not know where to find these entries in the mp4 file - if anybody does know something about this, please let me know :)

I guess it's in the VLC source code, but it's probably not going to be easy for me....

**Comment by darnn** [comment]
Clever ffmpeg-GUI manages to extract an SRT file with the following contents:

1
00:00:01,601 --> 00:00:06,639
<font face="Monospace">{\an7}\h\hparticularly did on this,
suggested that he thought that</font>

So I guess it's something, but it's not perfect. Slicing up the file into a tiny chunk might be exacerbating it, though.

**Comment by GrampaWildWilly** [comment]
Clever ffmpeg GUI.  Handy little tool.  Thanks for mentioning it.  I got portable version 3.1.1.  It extracted both of the subtitles, not just one:

-------------------------

1
00:00:01,601 --> 00:00:06,639
<font face="Monospace">{\an7}\h\hparticularly did on this,
suggested that he thought that</font>

2
00:00:06,672 --> 00:00:06,705
<font face="Monospace">{\an7}\h\h\hthere was one particular
component in the air that was</font>

-------------------------

I wish there were a way to tell it to extract vtt but there doesn't seem to be a way to do that.  In fact, this app doesn't seem to have any settings at all, just its functions.  I decided to run the error check function & got 16 lines of errors like this:

[null @ 0000000000555b00] Application provided invalid, non monotonically increasing dts to muxer in stream 0: 53 >= 53

Maybe that's a clue to what's going on with this.  In generating the little log report, it prompted me for where I have Notepad++.  Fortunately, I have that.  That was on the recommendation of some guy named Nik.  Perhaps you've heard of him?

I would really like to see the actual ffmpeg commands this app generates.  I was trying to extract the captions with ffmpeg manually but had no success.  This app shows I simply don't know enough to get it right.

**Comment by GrampaWildWilly** [comment]
Wait.  That post didn't show the complete subtitle lines.  Here.  Let me attach the actual file.  There is more to it than what you see in that post.  I had to adjust the file name because GitHub doesn't accept .srt as a file extension for attachments.

[Subtitle_S_extracted .srt.txt](https://github.com/SubtitleEdit/subtitleedit/files/11886131/Subtitle_S_extracted.srt.txt)

**Comment by Shake34** [comment]
Thank you so much for all your efforts!

I've tried extracting subtitles on this type of video using "Clever ffmpeg GUI" and it works great.

You don't know how grateful I am for your help! 
I've been trying to extract subtitles from these files for years.
**Thanks thanks thanks!**


I post a small example extracted from another file.

---------------------------
1
00:00:01,335 --> 00:00:10,177
<font face="Monospace">{\an7}[music playing]</font>

2
00:00:15,516 --> 00:00:17,618
<font face="Monospace">{\an7}At the end of the last lecture,</font>

3
00:00:17,651 --> 00:00:20,754
<font face="Monospace">{\an7}we left France on the edge of
\h\h\h\h\h\h\h\h\h\hdisaster.</font>

4
00:00:20,787 --> 00:00:22,989
<font face="Monospace">{\an7}\h\h\h\h\h\hThe king was
bankrupt and desperate,</font>

5
00:00:23,023 --> 00:00:25,792
<font face="Monospace">{\an7}his flabby back up
against the wall.</font>

6
00:00:25,826 --> 00:00:27,928
<font face="Monospace">{\an7}Louis XVI saw only one way out.</font>

7
00:00:27,961 --> 00:00:33,533
<font face="Monospace">{\an7}\h\h\hIn 1788 he agreed to
call the Estates-General.</font>

8
00:00:33,600 --> 00:00:36,870
<font face="Monospace">{\an7}\h\hThis assembly of elected
deputies would offer him the</font>

9
00:00:36,904 --> 00:00:39,540
<font face="Monospace">{\an7}\hopinions of the
people of France.</font>

---------

I could erase the words

**<font face="Monospace">{\an7}**

simply doing find and replace with Notepad+.

And the same with all those "\h"
I don't understand why sometimes there are many, sometimes few, sometimes none.


If you know a faster way to delete them, please let me know.

Thank you again!


edit: The line of each sub "font face="Monospace"" it doesn't show when I post the message.

**Comment by GrampaWildWilly** [comment]
If I understand things correctly, the {an#} things tell the video player where to position the subtitles on the screen.  There are things in the SE settings dialog that sort of document what the numbers mean.  It's upper left, upper center, upper right, and so on.

I had to look up what the \h things are.  I found this reference:

https://movian.tv/boards/15/topics/8105

Apparently, \h indicates a space that the author of the subtitles wants to make sure appears in the subtitle.

There's also the tags within angle brackets.  Those are a bit tricky to display here.  GitHub interprets them as HTML, even if what's inside them isn't particularly HTML.  Those are part of the instructions to the video player of how to display the captions.

Why do you want to remove them all?  After you remove all of that, what are you doing with what is left?
